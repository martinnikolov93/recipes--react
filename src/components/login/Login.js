import React from 'react';

import { Link } from 'react-router-dom'
import UserContext from '../../Context';

import { isEmptyObject } from '../../utils/helpers'

class Login extends React.Component {
    state = {
        email: '',
        emailErrors: false,
        password: '',
        passwordErrors: false,
    }

    emailValidator = (email) => {
        let emailErrors = {}

        if (!email) {
            emailErrors.required = 'Email is required.';
        }

        if (isEmptyObject(emailErrors)) {
            return false
        }

        return emailErrors
    }

    passwordValidator = (password) => {
        let passwordErrors = {}

        if (!password) {
            passwordErrors.required = 'Password is required.';
        }

        if (isEmptyObject(passwordErrors)) {
            return false
        }

        return passwordErrors
    }

    submitHandler = (event) => {
        event.preventDefault()

        let { email, password } = this.state

        const emailErrors = this.emailValidator(email)
        const passwordErrors = this.passwordValidator(password)

        this.setState({ emailErrors, passwordErrors })

        if (emailErrors || passwordErrors) {
            console.log('form wont be submited')
            return
        }

        fetch('http://localhost:9999/api/user/login', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                email,
                password
            })
        })
        .then((response) => {
            const user = response.json()
            const authToken = response.headers.get('auth-token')
            document.cookie = `auth-token=${authToken}`
            this.context.logIn(user)
            return user
        })
        .then((json) => console.log(json))

        console.log('form was submited')
    }

    changeHandlerEmail = (event) => {
        const email = event.target.value
        const emailErrors = this.emailValidator(email)
        this.setState({ email: event.target.value, emailErrors: emailErrors })
    }

    changeHandlerPassword = (event) => {
        const password = event.target.value
        const passwordErrors = this.passwordValidator(password)
        this.setState({ password: password, passwordErrors: passwordErrors })
    }

    render() {
        let { email, emailErrors, password, passwordErrors } = this.state

        return (
            <div>
                <form onSubmit={this.submitHandler}>
                    <div>
                        <label htmlFor='email'></label>
                        <input className={emailErrors ? "input-error" : ""} id='email' type="email" placeholder="Email" value={email} onChange={this.changeHandlerEmail} />
                        <div className="input-error-text">{emailErrors ? Object.values(emailErrors)[0] : null}</div>
                    </div>
                    <div>
                        <label htmlFor='password'></label>
                        <input className={passwordErrors ? "input-error" : ""} id='password' type="password" placeholder="Password" value={password} onChange={this.changeHandlerPassword} />
                        <div className="input-error-text">{passwordErrors ? Object.values(passwordErrors)[0] : null}</div>
                    </div>
                    <button>Login</button>
                </form>
                <br />
                <div>Don't have an account? Go to <Link to='/register'>Register Page</Link> and create one!</div>
            </div>
        )
    }
}

Login.contextType = UserContext;

export default Login