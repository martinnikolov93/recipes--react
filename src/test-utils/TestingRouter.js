import React from 'react';
import UserContext from '../Context';
import { BrowserRouter, Route } from 'react-router-dom'

const TestingRouter = ({user, children}) => {
    return (
        <BrowserRouter>
            <Route>
                <UserContext.Provider value={user}>
                    {children}
                </UserContext.Provider>
            </Route>
        </BrowserRouter>
    )
}

export default TestingRouter