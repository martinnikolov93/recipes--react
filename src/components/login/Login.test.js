import React from 'react';
import renderer from 'react-test-renderer'
import Login from './Login';
import TestingRouter from '../../test-utils/TestingRouter';

const userDummy = {
    loggedIn: false
}

describe('Login component', () => {
    it('Should render login component', () => {
        const tree = renderer.create(
            <TestingRouter user={userDummy}>
                <Login />
            </TestingRouter>
        ).toJSON()

        expect(tree).toMatchSnapshot()
    })
})