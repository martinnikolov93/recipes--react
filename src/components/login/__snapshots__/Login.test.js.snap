// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component Should render login component 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div>
      <label
        htmlFor="email"
      />
      <input
        className=""
        id="email"
        onChange={[Function]}
        placeholder="Email"
        type="email"
        value=""
      />
      <div
        className="input-error-text"
      />
    </div>
    <div>
      <label
        htmlFor="password"
      />
      <input
        className=""
        id="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
      <div
        className="input-error-text"
      />
    </div>
    <button>
      Login
    </button>
  </form>
  <br />
  <div>
    Don't have an account? Go to 
    <a
      href="/register"
      onClick={[Function]}
    >
      Register Page
    </a>
     and create one!
  </div>
</div>
`;
